wpical/src/main/native/cpp/fieldcalibration.cpp

```diff
@@ -0,0 +1,588 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "fieldcalibration.h"
+
+#include <algorithm>
+#include <filesystem>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "tag36h11.h"
+
+struct Pose {
+  Eigen::Vector3d p;
+  Eigen::Quaterniond q;
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+};
+
+struct Constraint {
+  int id_begin;
+  int id_end;
+  Pose t_begin_end;
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+};
+
+struct CameraModel {
+  Eigen::Matrix<double, 3, 3> intrinsic_matrix;
+  Eigen::Matrix<double, 8, 1> distortion_coefficients;
+};
+
+class PoseGraphError {
+ public:
+  explicit PoseGraphError(Pose t_ab_observed)
+      : m_t_ab_observed(std::move(t_ab_observed)) {}
+
+  template <typename T>
+  bool operator()(const T* const p_a_ptr, const T* const q_a_ptr,
+                  const T* const p_b_ptr, const T* const q_b_ptr,
+                  T* residuals_ptr) const {
+    // Tag A
+    Eigen::Map<const Eigen::Matrix<T, 3, 1>> p_a(p_a_ptr);
+    Eigen::Map<const Eigen::Quaternion<T>> q_a(q_a_ptr);
+
+    // Tag B
+    Eigen::Map<const Eigen::Matrix<T, 3, 1>> p_b(p_b_ptr);
+    Eigen::Map<const Eigen::Quaternion<T>> q_b(q_b_ptr);
+
+    // Rotation between tag A to tag B
+    Eigen::Quaternion<T> q_a_inverse = q_a.conjugate();
+    Eigen::Quaternion<T> q_ab_estimated = q_a_inverse * q_b;
+
+    // Displacement between tag A and tag B in tag A's frame
+    Eigen::Matrix<T, 3, 1> p_ab_estimated = q_a_inverse * (p_b - p_a);
+
+    // Error between the orientations
+    Eigen::Quaternion<T> delta_q =
+        m_t_ab_observed.q.template cast<T>() * q_ab_estimated.conjugate();
+
+    // Residuals
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
+    residuals.template block<3, 1>(0, 0) =
+        p_ab_estimated - m_t_ab_observed.p.template cast<T>();
+    residuals.template block<3, 1>(3, 0) = T(2.0) * delta_q.vec();
+
+    return true;
+  }
+
+  static ceres::CostFunction* Create(const Pose& t_ab_observed) {
+    return new ceres::AutoDiffCostFunction<PoseGraphError, 6, 3, 4, 3, 4>(
+        new PoseGraphError(t_ab_observed));
+  }
+
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+ private:
+  const Pose m_t_ab_observed;
+};
+
+static CameraModel load_camera_model(std::string path) {
```

Comment:
```suggestion
inline CameraModel load_camera_model(std::string path) {
```

Reply:
I mean, does it really matter? I feel like the compiler will sort it out

Reply:
Any particular reason for inline over static?

Reply:
Seems like a toss-up: https://stackoverflow.com/a/7767858

By the way, this function should take std::string_view instead of std::string to avoid a copy.

Reply:
any particular reason for inline over static?